swift/src/IceImpl/LocalObject.mm

```diff
@@ -38,16 +41,24 @@ + (nullable instancetype)getHandle:(std::shared_ptr<void>)cppObject
     }
     @synchronized([ICELocalObject class])
     {
-        std::unordered_map<void*, __weak ICELocalObject*>::const_iterator p = cachedObjects->find(cppObject.get());
+        auto p = cachedObjects->find(cppObject.get());
         if (p != cachedObjects->end())
         {
-            // assert(p->second);
-            return p->second;
-        }
-        else
-        {
-            return [[[self class] alloc] initWithCppObject:std::move(cppObject)];
+            // Get a strong reference to the object. If it's nil, preemptively remove it from the cache,
```

Comment:
I am not following the logic.

Say another thread is releasing the last-reference of the ObjC object, and is waiting to enter the synchronization in dealloc below to remove the entry in cachedObjects.
Meaning the entry is still in cacheObjects.

Now, p points to this "about to die" entry. So we remove the entry and create a new entry on line 61, and release the synchronization.

But now the other thread can run dealloc and remove the new entry we've just created... not good.

Reply:
This was essentially the previous issue. We could end up in the situation where dealloc was waiting for the mutex while this function is looking in the map. The new update gets the weak ptr the creates a a strong ptr. 

That's the point of `ICELocalObject* obj = p->second;` If we're in the situation objc is deallocating the object pointed to be p, then this assignment will return nil. Otherwise it will increment the ref count and there's no deallocation sorry.

Reply:
https://github.com/bernardnormier/ice/blob/pr/externl/3042/swift/src/IceImpl/LocalObject.mm shows the issue introduced by this PR, and a solution.